versionbits refactoring

ajtowns opened the pull request:
Increases the encapsulation/modularity of the versionbits code, moving more of the logic into the versionbits module rather than having it scattered across validation and rpc code. Updates unit/fuzz tests to test the actual code used rather than just a close approximation of it.

ajtowns committed:
versionbits: Use std::array instead of C-style arrays

ajtowns committed:
versionbits: Remove params from AbstractThresholdConditionChecker

For an abstract class, specifying parameters in detail serves no point;
and for the concrete implementation, changing the consensus parameters
between invocations doesn't make sense. So simplify the class by removing
the consensus params from the method arguments, and just make it a member
variable in the concrete object where needed. This also allows dropping
dummy parameters from the unit/fuzz tests.

ajtowns committed:
consensus/params: Move version bits period/threshold to bip9 param

Rather than having the rule change period/threshold be constant for all
potential deployments on a chain, have it be specific to the deployment
itself. This both matches history (BIP 9 specified a 2016 block period
and 1916 block threshold; BIP 91 specified a 336 block period and 269
block threshold; and BIP 341 specified a 2016 block period and 1815
block threshold), and allows the code to be simplified, as only the
BIP9Deployment structure is needed, not the full Consensus::Params
structure.

ajtowns committed:
versionbits: Move getdeploymentinfo logic to versionbits

Rather than having the RPC code have knowledge about how BIP9 is
implemented, create a reporting function in the versionbits code, and
limit the RPC code to coverting the result of that into Univalue/JSON.

ajtowns committed:
versionbits: Move BIP9 status logic for getblocktemplate to versionbits

Rather than having the RPC code have knowledge about how BIP9 is
implemented, create a reporting function in the versionbits code, and
limit the RPC code to coverting the result of that into the appropriate
output for getblocktemplate.

ajtowns committed:
versionbits: Simplify VersionBitsCache API

Replaces State() (which returned ACTIVE/STARTED/etc) with IsActiveAfter()
which just returns a bool, as this was all State was actually used
for. Drops Mask(), which was only used in tests and can be replaced with
`1<<bit`, and also drops StateSinceHeight() and Statistics(), which are
now only used internally for Info().

ajtowns committed:
versionbits: Split out internal details into impl header

ajtowns committed:
versionbits: Expose StateName function

Rather than essentially duplicating StateName in the unit tests, expose
it via the impl header.

ajtowns commented:
At the end of this sequence of patches the `VersionBitsCache` object provides a standalone interface to all the "versionbits" features needed by bitcoin core, which means that if the BIP9/speedy trial implementation is changed in future, it should be possible to do that in one place, without needing to change validation/RPC code as well.

This also simplifies and modernises some of the code, much of which has been proposed previously (eg, removing the params arguments from AbstractThresholdConditionChecker was included in the first versions of #21380; making threshold a per-deployment params was part of #19573 and #21392).

After these changes, it should be possible to simplify the unit/fuzz tests a little, as they should now be able to use `VersionBitsConditionChecker` directly, since it no longer depends on access to `Consensus::Params`.

naumenkogs commented:
Concept ACK. Looks cleaner. Will look in more detail when you undraft it.

luke-jr reviewed (commented):
Concept NACK, refactoring without a purpose, and probably makes more work for merging BIP8.

ajtowns committed:
versionbits: Expose VersionBitsConditionChecker via impl header

ajtowns committed:
versionbits: Change BIP9Stats to uint32_t types

ajtowns committed:
tests: refactor versionbits fuzz test

Test `VersionBitsConditionChecker` behaviour directly, rather than
reimplementing it, thus slightly improving fuzz test coverage of the
real code.

ajtowns committed:
tests: refactor versionbits unit test

Base the unit test directly on `VersionBitsConditionChecker`, slightly
improving coverage, in particular adding coverage for the the logic
regarding setting the TOP_BITS.

TheCharlatan commented:
Concept ACK

ajtowns commented:
Rebased

TheCharlatan commented:
> and probably makes more work for merging BIP8.

Is there a PR open for such logic?

TheCharlatan commented on src/versionbits_impl.h:
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2016-2022 The Bitcoin Core developers
```
In commit 49afd714b40115b15ea0ef146b48e9d7a403a596

Can you give some more motivation for splitting this out into its own header, especially since it seems to be introducing a new circular dependency? It also not immediately clear to me why a file called implementation only holds abstract declarations.  From what I understand this is supposed to hold declarations that are only relevant for external use by tests.

sedited reviewed (commented):
Reviewed up to the last commit. This looks very nice, I especially like the move of business logic out of the rpc into a more appropriate module. I also think the new place for `m_warning_caches` is much less confusing.

ajtowns commented on src/validation.cpp:
```diff
@@ -2378,8 +2378,20 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
-    int Period() const override { return m_chainman.GetConsensus().nMinerConfirmationWindow; }
-    int Threshold() const override { return m_chainman.GetConsensus().nRuleChangeActivationThreshold; }
+    int Period() const override {
+        if (m_chainman.GetParams().IsTestChain()) {
```
Changed it to set `period = chainparams.GetConsensus().DifficultyAdjustmentInterval();` which is much the same, but doesn't rely on the TESTDUMMY deployment.

ajtowns committed:
versionbits: Move WarningBits logic from validation to versionbits

darosior reviewed (commented):
Concept ACK. A couple comments as i was reading through, will delve deeper soon.

ajtowns commented on src/validation.cpp:
```diff
@@ -2378,8 +2378,20 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
+            return m_chainman.GetConsensus().DifficultyAdjustmentInterval();
+        } else {
+            return 2016;
+        }
```
For non-testnets, the 2016 and 1815 figures WarningBitsConditionChecker in versionbits.cpp (by the end of this PR) are constrained by what we might expect to see in future from consensus change coordination signalling; while that currently matches DifficultyAdjustmentInterval, it doesn't need to -- for instance BIP 91 had a window of 336 blocks, and I think in practice would not have (did not) trigger this warning code. So I think it makes sense to leave them as constants, though perhaps some in-line documentation would help?

darosior replied:
BIP9 does not define the period as configurable, it's always the retarget period. No strong opinion, but i find it surprising that the BIP9 warning mechanism should support (a specific type of) non-BIP9 deployments.

ajtowns commented on src/versionbits.cpp:
```diff
@@ -205,6 +217,35 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
+    const bool has_signal = (STARTED == current_state || LOCKED_IN == current_state);
+    if (has_signal) {
+        result.stats.emplace(Statistics(&block_index, params, id, &result.signalling_blocks));
+        if (LOCKED_IN == current_state) {
+            result.stats->threshold = 0;
+            result.stats->possible = false;
+        }
```
This is just clearing those members because they don't make sense in the LOCKED_IN state. The earlier line `result.current_state = StateName(current_state)` is what's reporting the state. Make sense?

darosior replied:
I'm not talking about the state but the stats. My comment was about how in this commit you replace `if (ThresholdState::LOCKED_IN != current_state)`  by `if (info.stats->threshold > 0 || info.stats->possible) {` to show the `threshold` and `possible` fields in the RPC, and this relies on these specific lines. I thought the only reason you had this logic here was to convey to the RPC whether to show these fields, which is a bit convoluted. I don't think this is worth touching at this point.

ajtowns replied:
> I thought the only reason you had this logic here was to convey to the RPC whether to show these fields, which is a bit convoluted. I don't think this is worth touching at this point.

Ah, no, it's two separate things in my head: the versionbits code quoted above is making sure the values returned in the structure are as sane as they can be, and the RPC code is taking a special case that wouldn't make sense (threshold=0 always implies possible=true if the signalling means anything) and trying to avoid confusing people. Just reporting it always and leaving it up to the caller to ignore the data would be possible, would make the statistics object not have any optional fields.

> I don't think this is worth touching at this point.

Okay, will mark as resolved.

TheCharlatan commented on src/versionbits.h:
```diff
@@ -43,15 +45,25 @@ typedef std::map<const CBlockIndex*, ThresholdState> ThresholdConditionCache;
+
+/** Detailed status of an enabled BIP9 deployment */
+struct BIP9Info {
+    int since{0};
```
Doc nit: Maybe add some docstrings here, just like currently done with BIP9Stats?

ajtowns replied:
Added.

sedited reviewed (approved):
ACK 7565563bc7a5bb98ebf03a7d6881912a74d3f302

ajtowns committed:
versionbits: docstrings for BIP9Info

ajtowns committed:
test: add IsActiveAfter tests for versionbits

ajtowns commented on src/test/versionbits_tests.cpp:
```diff
@@ -16,39 +16,41 @@
 /* Define a virtual block time, one block per 10 minutes after Nov 14 2014, 0:55:36am */
 static int32_t TestTime(int nHeight) { return 1415926536 + 600 * nHeight; }
 
```
Added a bunch of checks to the unit test. The fuzz tester checks at a lower level (`ThresholdConditionCache` vs `VersionBitsCache`) so that's not really suitable.

sedited reviewed (approved):
Re-ACK e3014017bacff42d8d69f3061ce1ee621aaa450a

Just addressed nits; added some more docs and tests for `IsActiveAfter`.

darosior reviewed (approved):
ACK e3014017bacff42d8d69f3061ce1ee621aaa450a

achow101 commented:
ACK e3014017bacff42d8d69f3061ce1ee621aaa450a

Sjors reviewed (commented):
Post merge ACK (did not study the test and fuzz changes). Agree this was a good time to merge.

The changes in 1198e7d2fd665bf2bc49fd26773d4fd5fbc2b716 which move logic from RPC to `versionbits.cpp` should make it easier to implement version bit signalling support in Stratum v2 (IPC). And reminded me that I actually need to look into that.

3bd32c20550e69688a4ff02409fb34b9a637b9c4 would have been easier to follow if had been split between one commit that moves and another that modified it (adding `m_caches`, `period`, etc). But perhaps this was hard to avoid due to losing access to `m_chainman`.

I ran a partial IBD with `-assumevalid=0` to sanity check that Taproot still activates at height 709,632.

Jewellorraine commented:
@jewel Lorraine
